<a href="https://colab.research.google.com/github/Devika-Aravind/ML_intern_Dvka/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Devika-Aravind/ML_intern_Dvka/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

I want to build a system that looks at all the pages and hands the content team a ranked list, so they know which pages to check first for a refresh.
I picked this lane because in Week 1, I already practiced building a hand-written rule and comparing it against a model using Precision@K. This lane builds directly on that same idea, just at a bigger scale.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

So Decision : My work helps to decide which pages to be reviewed first for a refresh.    
Action: Content team/ content updaters, they take my ranked list and go fix the top pages.

Cost of a Wrong call: If my list flags a page that didn't need fixing, the content team wastes time by checking it for nothing. But if my list misses a page that really need fixing, that page keep losing traffic and visibility while staying stale and unnoticed.

In [9]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [10]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import os
import sys
import subprocess
import pandas as pd

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/Devika-Aravind/ML_intern_Dvka"
REPO_DIR = "ML_intern_Dvka"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)

print("Working dir:", os.getcwd())
assert os.path.exists("data/raw/content_refresh_anonymized.csv"), "CSV not found — check repo cloned correctly"

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

# 1. How many pages are stale AND still visible? (candidates for refresh)
stale_visible = ((df["days_since_last_update"] >= 180) & (df["impressions_90d"] >= 500)).sum()
print(f"Stale + visible pages (refresh candidates): {stale_visible} out of {len(df)}")

# 2. What percent of pages are declining?
declining_pct = (df["trend_direction"] == "down").mean() * 100
print(f"Percent of pages declining: {declining_pct:.1f}%")

# 3. Average impressions for declining vs growing pages
avg_impressions = df.groupby("trend_direction")["impressions_90d"].mean()
print("\nAverage impressions by trend direction:")
print(avg_impressions.round(0))



Working dir: /ML_intern_Dvka/ML_intern_Dvka/ML_intern_Dvka
Stale + visible pages (refresh candidates): 17 out of 30000
Percent of pages declining: 54.2%

Average impressions by trend direction:
trend_direction
down      4919.0
flat        21.0
new        164.0
stable    9214.0
up        4716.0
Name: impressions_90d, dtype: float64


Out of 30,000 pages, only 17 are both stale and highly visible. A narrow,edge-case slice. But the the bigger signal is that 54.2% of all pages are currently declining, and these declining pages still average nearly 4,919 impressions each - meaning they still have real traffic worth protecting. This broader declining signal is what makes this lane worth pursuing for the next 7 weeks.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

My ranking is a list based on data and obervation - a best guess, not a guranteed fact.
Even if a page gets better after being reviewed i can't be sure that my ranking or my system is what caused it. (Other things like the page becoming trendy again, or unrelated changes could explain the improvement too)

I also can't say that my system knows how Google's algorithm actually works, i only have search behaviour data (impressions,clicks,position), not the algorithm itself.

In [11]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.